SAA Perks & Benefits Protected Content

sherakama opened the pull request:
# READY FOR REVIEW

# Summary
- Perks for members work

# Review By (Date)
- ?

# Criticality
- Medium

# Review Tasks
- Review Editor Experience: http://app.storyblok.com/#!/me/spaces/147533/stories/0/0/122607995/blok/062a1947-bc46-47ab-8c88-3160255e979b?s=1
- Review User Experience: https://deploy-preview-329--stanford-alumni.netlify.app/test/perks-auth-test/
- Check mobile versions
- Validate data and security by logging in as different users

sherakama commented on netlify.toml:
```diff
@@ -67,21 +62,27 @@
   # Path to the sitemap URL (optional, default = /sitemap.xml)
   sitemapPath = "/sitemap/sitemap-index.xml"
+
+## ########################## ##
+## REMOVE BEFORE MERGE TO DEV ##
+## ########################## ##
+[build.environment]
+  VAULT_OVERWRITE = "true"
```
```suggestion
```

sherakama commented on src/api/auth/profile.js:
```diff
@@ -0,0 +1,36 @@
+  const mp = new MegaProfile();
+  try {
+    const { user } = req;
+    const { data: contact } = await mp.get(
+      `/${req.user.encodedSUID}/profiles/contact`
+    );
```
This endpoint won't be ready for when we plan to go live with this. We are going to have to refactor this code to fetch from the `/fullgg` endpoint. 


```suggestion
```

sherakama replied:
@nplowman 

Some bad news for you on this endpoint. Unfortunately, it won't be available in production for when we want to go live with this work. Please refactor to connect to the `/fullgg` endpoint. This is the same endpoint used here https://github.com/SU-SWS/ood_online_giving/blob/dev/src/api/auth/profile/%5Bsuid%5D.js

You will need to maintain the `/affiliations` endpoint however as the `/fullgg` endpoint does not have the information.

sherakama commented on src/api/auth/profile.js:
```diff
@@ -0,0 +1,36 @@
+    );
+    const mpUser = { user, contact, affiliations: affiliations.affiliations };
+    return res.status(200).json(mpUser);
+    // TODO: ADAPT-4438 Once we have the user data, we can append the megaprofile data.
+    // const { addresses } = await mp.get(`${user.encodedSUID}/profiles/addresses`);
+    // const { emails } = await mp.get(`${user.encodedSUID}/profiles/emails`);
+    // const { phoneNumbers } = await mp.get(`${user.encodedSUID}/profiles/phonenumbers`);
+    // ...additional endpoints here
+    // const mpUser = { user, contact, addresses, emails, phoneNumbers, ...additional endpoints here };
+    // res.status(200).json(mpUser);
```
```suggestion
```

sherakama commented on src/api/private-proxy.js:
```diff
@@ -0,0 +1,45 @@
+    `/${req.user.encodedSUID}/profiles/affiliations`
+  );
+
+  // TODO: Check for JWT, decode it and confirm the user is authorized.
```
This isn't todo any more is it?

nplowman replied:
I belive I removed this in my refactoring PR.

sherakama commented on src/components/components.js:
```diff
@@ -22,6 +22,7 @@ import { SBGlobalHeader } from './storyblok/globalHeader';
 import { SBGlobalHeaderPicker } from './storyblok/globalHeaderPicker';
 import { SBGrid } from './storyblok/grid';
 import { SBHeading } from './storyblok/heading';
+import GiveGabForm from './embed/giveGabForm';
```
We can remove this from this PR.

sherakama commented on src/components/components.js:
```diff
@@ -86,6 +91,7 @@ const ComponentList = {
   embedVideo: EmbedVideo,
   event: Event,
   eventCard: EventCard,
+  giveGabForm: GiveGabForm,
```
```suggestion
```

sherakama commented on src/components/embed/dynaScript.js:
```diff
@@ -0,0 +1,59 @@
+/* eslint-disable no-alert */
```
Delete this file.

sherakama commented on src/components/embed/giveGabForm.js:
```diff
@@ -0,0 +1,94 @@
+/* eslint-disable camelcase */
```
delete this file.

sherakama commented on src/pages/protected-page.js:
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
```
This file can be deleted.

sherakama commented on src/utilities/MegaProfile.js:
```diff
@@ -0,0 +1,57 @@
+  get = async (url, config = {}) =>
+    this.request({ ...config, method: 'GET', url });
+
+  // TODO: Remove this method once test-405 is confirmed?
+  post = async (url, data, config = {}) =>
+    this.request({
+      ...config,
+      method: 'POST',
+      url,
+      data,
+    });
```
No post is needed here. This can be removed.

sherakama commented on src/utilities/giveGabVars.js:
```diff
@@ -0,0 +1,2 @@
+// TODO: Create global javascript (window) variables for giveGabForm. (ADAPT-4441)
```
This file can be removed.

sherakama commented on src/utilities/mocks/index.js:
```diff
@@ -0,0 +1,2 @@
+export * from './contact';
```
This will need refactoring.

yvonnetangsu commented:
@sherakama It sounds like there's a release coming out soon for Alumni? I have a ticket to do some a11y review for the new components that are getting added. Do you have a list that I need to look at? I know we have the new masthead/user menu with the login, are there some perk/discount code hide/show component as well? 😃

sherakama commented:
@yvonnetangsu,

The work in this PR is not slated to go out for a little while yet but you can use this diff to identify all the things that have changed. The steps to test and example pages can be found in the PR comments here: https://github.com/SU-SWS/saa_alumni/pull/325

https://deploy-preview-329--stanford-alumni.netlify.app/test/perks-auth-test

Please use the preview deploy site for your A11y review with SODA

sherakama commented on src/components/navigation/accountLinks.js:
```diff
@@ -0,0 +1,150 @@
+                    url: 'https://alumni.stanford.edu/get/page/my-account/profile',
+                    linktype: 'url',
+                  },
+                  text: 'My Account',
```
This link is appearing as a duplicate in the preview build. I suspect this is a storyblok change?

yvonnetangsu replied:
Even if the duplicate link is removed, it looks like some responsive tweak is needed here
![Screen Shot 2022-04-26 at 1 34 22 PM](https://user-images.githubusercontent.com/42749717/165387720-794dd2bc-039c-4964-a319-e661ff54f14d.png)

yvonnetangsu replied:
Thanks, @sherakama !

yvonnetangsu replied:
@sherakama Is this still the link to look at? It seems like there's still 2 "My Account" links in the utility nav. 
https://deploy-preview-329--stanford-alumni.netlify.app/test/perks-auth-test/

We might still need some responsive styling tweaks even if we remove one of the duplicate links though. For example, this is for 375px, to squeeze everything in, there's very little space between the links. Perhaps we could ask Kerri if it's ok for change "Alumni Directory" to "Directory"? That would buy us some spacing 👀 
![Screen Shot 2022-04-29 at 3 00 32 PM](https://user-images.githubusercontent.com/42749717/166074844-3b60fefe-b703-4486-9461-1b8949c85d0b.png)

yvonnetangsu replied:
I also seem to have trouble logging in at the moment. I remember using `zguan` as the login before, so I tried using that account. It was processing at first, but didn't end up logging me in and there was no login error displayed either 😬

sherakama replied:
Yeah, let me circle back on this. It looks like the affiliations endpoint is still down...

yvonnetangsu commented:
@sherakama I'm now able to login so I did an initial review before sending it to SODA. Not sure who worked on the components so I'll just leave my feedback here:

1. The "Copy promo code" CTA link could use some hover/focus styles. I believe it can actually use the exact same component/variants as the CTA links in the local footer, but with the leading FA icon enabled
![Screen Shot 2022-05-09 at 11 27 15 AM](https://user-images.githubusercontent.com/42749717/167473640-6239fa79-7e30-410d-b8c7-c4c2471cca22.png)
2. It seems like the change of content inside the Perk (from login CTA to "checking access" to displaying the content) would require some `aria-live` and related attributes to announce to the screen reader that the content has been updated.
3. I suspect that we'll need some focus management as well, but that could wait till we get recommendation from SODA
4. I feel like the content could use some padding (around the content, and also some space next to the "Copied" text)
Note: looks like the mock up for this is having some issues - will talk to design team about this
![Screen Shot 2022-05-09 at 11 23 57 AM](https://user-images.githubusercontent.com/42749717/167473296-f56bdb7f-bc39-41a0-9a48-9f9de31da9ed.png)
5. I saw this in the mock for the login prompt content. Are we still doing this? It seems different from our current version which only has a simple line of text "Log in to access special discounts for Members!"
![Screen Shot 2022-05-09 at 11 35 12 AM](https://user-images.githubusercontent.com/42749717/167475151-2ec504e1-81cd-447a-bb1a-d8ad74301a74.png)

yvonnetangsu commented:
@sherakama I also discussed with Kerri about the mobile styles for the utility nav. There are currently 2 options
1. Change all link font sizes to 13px, Remove the word "Check" in front of "Check email"
2. Remove the "My Account" link from the utility nav since it's inside the user menu already.
Kerri, Alison, and Dena mentioned that this would be a decision that Michael has to make.

sherakama commented:
@yvonnetangsu,

I was merging with the latest from dev and ended up updating the packages and upgraded to decanter-rc.0 in the process. Things still look tight to me but I will trust your keen eye over mine any day. Please have a look to ensure that nothing shifted out of whack.

The My Account link is now gone on mobile and only shows up on the desktop logged in state

I removed the work `check` from `check email` so that the utility menu fits on mobile portrait mode

I updated the text in the logged out state for the promo code. This content is all from a rich text field in storyblok: https://app.storyblok.com/#!/me/spaces/147533/stories/0/0/122607995/blok/062a1947-bc46-47ab-8c88-3160255e979b?s=1

To get the button to live in there I had to add it as a component to the wysiwyg options and rich text renderer. Please have a look and see the editor's experience in Storyblok. I only added the one CTA Button Link component so as not to open pandora's box on this. 

I'll see about adding an aria-live attribute to the containing wrapper.

yvonnetangsu commented:
> 

Thank you for this work, @sherakama ! I'll take a good look tomorrow. Want to make sure the Decanter rc update doesn't change anything 😄

yvonnetangsu commented:
Ooooo there is definitely some differences. Should I make any TW related changes in this PR?

sherakama commented:
> Ooooo there is definitely some differences. Should I make any TW related changes in this PR?

Yeah, go ahead. Looks like the children combinators still need work. Ie: `children:first`, `children:last`, `children:children`, etc

yvonnetangsu commented:
@sherakama I think it would need some find/replace and inspection (probably some of the group-hocus: classes, as that was from the 3rd party plugin and no easy way to write a 1 liner plugin). I'd be happy to help with that 😄

sherakama commented:
> @sherakama I think it would need some find/replace and inspection (probably some of the group-hocus: classes, as that was from the 3rd party plugin and no easy way to write a 1 liner plugin). I'd be happy to help with that 😄

Hmm yeah. Going to come back at this tomorrow. We have some wild classes to look at. eg: `last:children:children:first:children:last:su-hidden`

yvonnetangsu commented:
> > @sherakama I think it would need some find/replace and inspection (probably some of the group-hocus: classes, as that was from the 3rd party plugin and no easy way to write a 1 liner plugin). I'd be happy to help with that 😄
> 
> Hmm yeah. Going to come back at this tomorrow. We have some wild classes to look at. eg: `last:children:children:first:children:last:su-hidden`

LOL - that's probably me. I'm currently fixing some of the group-hocus ones. Things are looking better already.

yvonnetangsu commented:
@sherakama Just pushed up the first batch. Will do a more detail look for the more usual ones tomorrow or this week 😄

sherakama commented:
> probably some of the group-hocus: classes, as that was from the 3rd party plugin and no easy way to write a 1 liner plugin

How about:
`addVariant('group-hocus', [':merge(.group):focus &', ':merge(.group):hover &'])`

Hover over the list items in: https://play.tailwindcss.com/6yMmMtyeiU

yvonnetangsu commented:
> > probably some of the group-hocus: classes, as that was from the 3rd party plugin and no easy way to write a 1 liner plugin
> 
> How about: `addVariant('group-hocus', [':merge(.group):focus &', ':merge(.group):hover &'])`
> 
> Hover over the list items in: https://play.tailwindcss.com/6yMmMtyeiU

Your way is better than mine! Just approved your #343  🚀  Please merge when you're ready (not sure if you want to add anything else). The Itinerary crazy classes should work now. Let me do another inspection and see if we have any more of these variant combinations.

yvonnetangsu commented:
@sherakama When do you need this merged by? I want to build a full test page that more closely resemble the mock for a perk, e.g., with the Black background page and all.

sherakama commented:
Sometime before May 30. We are waiting on a change to the OAuth service for the megaprofile before we go live. So this isn't a rush but I want to be ready for when the API is up and out.

yvonnetangsu commented:
> Sometime before May 30. We are waiting on a change to the OAuth service for the megaprofile before we go live. So this isn't a rush but I want to be ready for when the API is up and out.

Oh good to know. I'll come back to this next week after the self evals are in then, and also we might hear back from Bhavika by then.

yvonnetangsu commented:
@sherakama New perk test page here:
https://perks--stanford-alumni.netlify.app/test/dark-perk-auth-test/
I use the Dark Background Page content type instead of the Basic Page. Since we aren't building a custom Perk page content type anymore, this gets it a bit closer to the mockup. I'll send this over to SODA, and will continue to look for anything left to tweak.

sherakama commented:
You are a 🪨 ⭐ ! Thank you @yvonnetangsu

yvonnetangsu commented:
> You are a 🪨 ⭐ ! Thank you @yvonnetangsu

A few more changes coming 😃  Thank you for putting this together, @sherakama !

yvonnetangsu commented on src/components/cards/promoCodeBanner.js:
```diff
@@ -1,63 +1,38 @@
+    <div className="su-type-1 su-rs-mb-0 su-text-center su-text-black-20">
+      {blok.promoCode}
+    </div>
+    <CopyButton
+      variant="link"
+      className="su-rs-mb-4"
+      copyText={blok.promoCode}
+      copySuccess={
+        <span>
+          <HeroIcon
+            iconType="check"
+            className="su-text-[3rem] su-inline-block su-text-palo-verde su-mr-03em"
+          />
+          Promo code copied!
+        </span>
+      }
+    >
+      <HeroIcon
+        iconType="document-duplicate"
+        className="su-text-[3rem] su-inline-block su-text-black-80 su-mr-03em group-hocus:su-text-white"
+      />
+      Copy promo code
+    </CopyButton>
```
@sherakama Here I'm using our existing component `CopyButton` instead of creating the functionality from scratch. I added this particular style (with some minor tweaks) as a variant of `CopyButton`. The original button variant is used on Trip pages, e.g., in the Overview section right sidebar here.
https://perks--stanford-alumni.netlify.app/travel-study/trips/botswana-south-africa-and-zambia/

yvonnetangsu commented:
@sherakama Added testing info for SODA in the ticket 🤞🏼 
https://stanfordits.atlassian.net/browse/ADAPT-4865

yvonnetangsu commented:
@sherakama Heard back from SODA - they only mentioned 2 issues. I think this is in great shape.

1. red arrow color not enough contrast inside the user menu (I replied asking if it's ok if the arrows are purely decorative here)
2. Can be fixed Storyblok side. Apparently there's text in the optional screen reader text "log in" in the Perk's log in button, so it says "log in" twice - easily fixed by removing that in Storyblok.

yvonnetangsu commented on src/components/navigation/accountLinks.js:
```diff
@@ -107,7 +107,7 @@ const AccountLinks = ({ mainLinkClasses }) => {
               </button>
               <ul
-                className={`su-transform-gpu su-transition su-origin-top md:su-origin-top-right su-bg-digital-red-dark su-z-10 su-list-none su-absolute su-rs-px-1 su-rs-pt-0 su-rs-pb-1 children:su-mb-02em su-w-screen su-mr-[-2rem] sm:su-mr-[-3rem] md:su-w-[32rem] su-right-0 su-text-left
+                className={`su-transform-gpu su-transition su-origin-top md:su-origin-top-right su-bg-cardinal-red-xdark su-z-10 su-list-none su-absolute su-rs-px-1 su-rs-pt-0 su-rs-pb-1 children:su-mb-02em su-w-screen su-mr-[-2rem] sm:su-mr-[-3rem] md:su-w-[32rem] su-right-0 su-text-left
```
Just made the change of background color in the user menu from `su-bg-digital-red-dark` to `su-bg-cardinal-red-xdark` (this is the actual color we want and pass contrast test). They were so similar I didn't even notice that it wasn't the right color before. We're all good with a11y now 😄

sherakama replied:
Woot! Thanks for handling that @yvonnetangsu

sherakama commented on netlify.toml:
```diff
@@ -62,16 +62,22 @@
   # Path to the sitemap URL (optional, default = /sitemap.xml)
   sitemapPath = "/sitemap/sitemap-index.xml"
+
+## ########################## ##
+## REMOVE BEFORE MERGE TO DEV ##
+## ########################## ##
+[build.environment]
+  VAULT_OVERWRITE = "true"
```
Remove before merge.

sherakama commented on netlify.toml:
```diff
@@ -46,8 +45,9 @@
-    secrets = ['secret/data/projects/adapt/certs', 'secret/data/projects/adapt/alumni']
+    # Order doesn't matter here as these will be fetched asyncronously.
+    # Try not to have dupes.
+    secrets = ['secret/data/projects/adapt/saml', 'secret/data/projects/adapt/perks']
```
This needs to change back to `/alumni`

sherakama reviewed (commented):
GTG

sherakama commented:
Here we go. Storyblok components have been synced to Dev and Vault is up-to-date.
